### ISSUE TITLE
Error: Message payload size does not match buffer size!

### ISSUE THREAD

**Issue by mnaamani** [Joystream/storage-node-joystream#51]
Getting this error frequently on our storage server. After node restarts uploads don't seem to work and requires to restart node one more time.

```
Apr 20 21:40:48 localhost node[13026]: events.js:174
Apr 20 21:40:48 localhost node[13026]:       throw er; // Unhandled 'error' event
Apr 20 21:40:48 localhost node[13026]:       ^
Apr 20 21:40:48 localhost node[13026]: Error: Message payload size does not match buffer size!
Apr 20 21:40:48 localhost node[13026]:     at parse (/home/joystream/storage-node-joystream/lib/joystream/protocols/envelope.js:43
Apr 20 21:40:48 localhost node[13026]:     at Deserializer._transform (/home/joystream/storage-node-joystream/lib/joystream/protoc
Apr 20 21:40:48 localhost node[13026]:     at Deserializer.Transform._read (_stream_transform.js:190:10)
Apr 20 21:40:48 localhost node[13026]:     at Deserializer.Transform._write (_stream_transform.js:178:12)
Apr 20 21:40:48 localhost node[13026]:     at doWrite (_stream_writable.js:410:12)
Apr 20 21:40:48 localhost node[13026]:     at writeOrBuffer (_stream_writable.js:394:5)
Apr 20 21:40:48 localhost node[13026]:     at Deserializer.Writable.write (_stream_writable.js:294:11)
Apr 20 21:40:48 localhost node[13026]:     at Socket.ondata (_stream_readable.js:689:20)
Apr 20 21:40:48 localhost node[13026]:     at Socket.emit (events.js:189:13)
Apr 20 21:40:48 localhost node[13026]:     at addChunk (_stream_readable.js:284:12)
Apr 20 21:40:48 localhost node[13026]: Emitted 'error' event at:
Apr 20 21:40:48 localhost node[13026]:     at Deserializer.onerror (_stream_readable.js:713:12)
Apr 20 21:40:48 localhost node[13026]:     at Deserializer.emit (events.js:189:13)
Apr 20 21:40:48 localhost node[13026]:     at onwriteError (_stream_writable.js:425:12)
Apr 20 21:40:48 localhost node[13026]:     at onwrite (_stream_writable.js:456:5)
Apr 20 21:40:48 localhost node[13026]:     at Deserializer.afterTransform (_stream_transform.js:94:3)
Apr 20 21:40:48 localhost node[13026]:     at Deserializer._transform (/home/joystream/storage-node-joystream/lib/joystream/protoc
Apr 20 21:40:48 localhost node[13026]:     at Deserializer.Transform._read (_stream_transform.js:190:10)
Apr 20 21:40:48 localhost node[13026]:     [... lines matching original stack trace ...]
Apr 20 21:40:48 localhost node[13026]:     at writeOrBuffer (_stream_writable.js:394:5)
Apr 20 21:40:48 localhost systemd[1]: storage-node.service: Main process exited, code=exited, status=1/FAILURE
Apr 20 21:40:48 localhost systemd[1]: storage-node.service: Failed with result 'exit-code'.
Apr 20 21:40:48 localhost systemd[1]: storage-node.service: Service hold-off time over, scheduling restart.
Apr 20 21:40:48 localhost systemd[1]: storage-node.service: Scheduled restart job, restart counter is at 2.
```

**Comment by jfinkhaeuser** [comment]
Thanks! Part of this is to do with the port probing the DHT does, but I'm not sure it's all there is. I'll investigate.

**Comment by mnaamani** [comment]
I'm seeing this before the error:
```
Apr 22 17:26:05 localhost node[32217]: Mon, 22 Apr 2019 17:26:05 GMT joystream::protocols:mutual Initialized 023931246c995d94523f1
Apr 22 17:26:05 localhost node[32217]: Mon, 22 Apr 2019 17:26:05 GMT joystream::protocols:multi Protocol found MutualAuthenticator
Apr 22 17:26:05 localhost node[32217]: Mon, 22 Apr 2019 17:26:05 GMT joystream::protocols:multi Mapping message range [ 0, 12 ]
Apr 22 17:26:05 localhost node[32217]: Mon, 22 Apr 2019 17:26:05 GMT joystream::protocols:multi Protocol found SyncProtocol
Apr 22 17:26:05 localhost node[32217]: Mon, 22 Apr 2019 17:26:05 GMT joystream::protocols:multi Mapping message range [ 0, 13 ]
Apr 22 17:26:05 localhost node[32217]: Mon, 22 Apr 2019 17:26:05 GMT joystream::protocols:multi Mapping is { '0': [ 0, 0 ],
Apr 22 17:26:05 localhost node[32217]:   '1': [ 0, 0 ],
Apr 22 17:26:05 localhost node[32217]:   '2': [ 0, 0 ],
Apr 22 17:26:05 localhost node[32217]:   '3': [ 0, 0 ],
Apr 22 17:26:05 localhost node[32217]:   '4': [ 0, 0 ],
Apr 22 17:26:05 localhost node[32217]:   '5': [ 0, 0 ],
Apr 22 17:26:05 localhost node[32217]:   '6': [ 0, 0 ],
Apr 22 17:26:05 localhost node[32217]:   '7': [ 0, 0 ],
Apr 22 17:26:05 localhost node[32217]:   '8': [ 0, 0 ],
Apr 22 17:26:05 localhost node[32217]:   '9': [ 0, 0 ],
Apr 22 17:26:05 localhost node[32217]:   '10': [ 0, 0 ],
Apr 22 17:26:05 localhost node[32217]:   '11': [ 0, 0 ],
Apr 22 17:26:05 localhost node[32217]:   '12': [ 0, 0 ],
Apr 22 17:26:05 localhost node[32217]:   '13': [ 13, 1 ],
Apr 22 17:26:05 localhost node[32217]:   '14': [ 13, 1 ],
Apr 22 17:26:05 localhost node[32217]:   '15': [ 13, 1 ],
Apr 22 17:26:05 localhost node[32217]:   '16': [ 13, 1 ],
Apr 22 17:26:05 localhost node[32217]:   '17': [ 13, 1 ],
Apr 22 17:26:05 localhost node[32217]:   '18': [ 13, 1 ],
Apr 22 17:26:05 localhost node[32217]:   '19': [ 13, 1 ],
Apr 22 17:26:05 localhost node[32217]:   '20': [ 13, 1 ],
Apr 22 17:26:05 localhost node[32217]:   '21': [ 13, 1 ],
Apr 22 17:26:05 localhost node[32217]:   '22': [ 13, 1 ],
Apr 22 17:26:05 localhost node[32217]:   '23': [ 13, 1 ],
Apr 22 17:26:05 localhost node[32217]:   '24': [ 13, 1 ],
Apr 22 17:26:05 localhost node[32217]:   '25': [ 13, 1 ],
Apr 22 17:26:05 localhost node[32217]:   '26': [ 13, 1 ] } { '0': 0, '1': 13 }
Apr 22 17:26:05 localhost node[32217]: Mon, 22 Apr 2019 17:26:05 GMT joystream::protocols:multi Registered incoming handler for 2
Apr 22 17:26:05 localhost node[32217]: Mon, 22 Apr 2019 17:26:05 GMT joystream::protocols:multi Handlers now { '2': { incoming: [ 
Apr 22 17:26:05 localhost node[32217]: Mon, 22 Apr 2019 17:26:05 GMT joystream:util:lru About to prune; have 1 and capacity is 100
Apr 22 17:26:05 localhost node[32217]: Mon, 22 Apr 2019 17:26:05 GMT joystream:util:lru Sorted keys are: [ [ 'IPv6:[::ffff:46.17.4
Apr 22 17:26:05 localhost node[32217]: Mon, 22 Apr 2019 17:26:05 GMT joystream:util:lru Have to prune -99 items.
Apr 22 17:26:05 localhost node[32217]: Mon, 22 Apr 2019 17:26:05 GMT joystream:util:lru Size after pruning 1
Apr 22 17:26:05 localhost node[32217]: Mon, 22 Apr 2019 17:26:05 GMT joystream::protocols:envelope Deserializer got <Buffer 47 45 
Apr 22 17:26:05 localhost node[32217]: events.js:174
Apr 22 17:26:05 localhost node[32217]:       throw er; // Unhandled 'error' event
Apr 22 17:26:05 localhost node[32217]:       ^
Apr 22 17:26:05 localhost node[32217]: Error: Message payload size does not match buffer size!
```

**Comment by mnaamani** [comment]
Then after the node starts again it crashes again with same error during startup:
```
Apr 22 17:26:05 localhost systemd[1]: Started Joystream Storage Node.
Apr 22 17:26:06 localhost node[5532]: Mon, 22 Apr 2019 17:26:06 GMT joystream:cli argv { storage: '/home/joystream/STORAGE' }
Apr 22 17:26:06 localhost node[5532]: Mon, 22 Apr 2019 17:26:06 GMT joystream:cli Configstore {
Apr 22 17:26:06 localhost node[5532]:   path: '/home/joystream/.config/configstore/joystream-storage.json' }
Apr 22 17:26:06 localhost node[5532]: Mon, 22 Apr 2019 17:26:06 GMT joystream:substrate:assets Init
Apr 22 17:26:06 localhost node[5532]: Mon, 22 Apr 2019 17:26:06 GMT joystream:substrate:roles Init
Apr 22 17:26:06 localhost node[5532]: Mon, 22 Apr 2019 17:26:06 GMT joystream:substrate:balances Init
Apr 22 17:26:06 localhost node[5532]: Mon, 22 Apr 2019 17:26:06 GMT joystream:substrate:identities Init
Apr 22 17:26:06 localhost node[5532]: Mon, 22 Apr 2019 17:26:06 GMT joystream:substrate:base Init
Apr 22 17:26:07 localhost node[5532]: Mon, 22 Apr 2019 17:26:07 GMT joystream:substrate:identities Initializing key from /home/joystream/st
Apr 22 17:26:07 localhost node[5532]: Mon, 22 Apr 2019 17:26:07 GMT joystream:substrate:identities Successfully initialized with address 5E
Apr 22 17:26:07 localhost node[5532]: Waiting for the account to be staked as a storage provider role...
Apr 22 17:26:07 localhost node[5532]: Staked, proceeding.
Apr 22 17:26:07 localhost node[5532]: Mon, 22 Apr 2019 17:26:07 GMT joystream:storage Initialized storage 265292ca-e7c3-4676-98b7-a47f6814b
lines 1585-1628
Apr 22 17:26:05 localhost node[32217]: Mon, 22 Apr 2019 17:26:05 GMT joystream::protocols:envelope Deserializer got <Buffer 47 45 54 20 2f 2
Apr 22 17:26:05 localhost node[32217]: events.js:174
Apr 22 17:26:05 localhost node[32217]:       throw er; // Unhandled 'error' event
Apr 22 17:26:05 localhost node[32217]:       ^
Apr 22 17:26:05 localhost node[32217]: Error: Message payload size does not match buffer size!
```

**Comment by jfinkhaeuser** [comment]
So far, the only success I've had is to isolate this in a test case. Which is sort of the hardest part, so I'm actually quite happy. To be continued.

**Comment by jfinkhaeuser** [comment]
Well, I do understand what is happening here now. It's actually very logically a bug, and the fix isn't too hard. But it's something that didn't crop up so easily during testing.

**Comment by jfinkhaeuser** [comment]
Well, this was a couple of bugs so far. And one more thing to change tomorrow, I think. The test cases were far too naive.

The TL;DR is that the test cases sent too few data packets, and did not yet use the network, but forwarded everything in-process from one stream instance to the other.

As a result, it never happened that two or more messages got concatenated into one network packet and had to be separated again. That's where the error message stems from, finding a buffer that does not match the message size because it contains three messages.

Another result was that events were very synchronized, which meant the two synchronization parties never talked over each other. That caused another number of issues. Now one party calmly waits for the other to finish. What's missing is that it then needs to respond in kind.

Lastly, the test cases did not really consider a synchronization stream to be bi-directional, but figured the full repository would only produce data, and the mirror would only consume data.

So, still a little ways to go, but not too much any longer.

**Comment by jfinkhaeuser** [comment]
This is the issue that keeps on giving.

Turns out there's a weird issue with hyperdrive which I couldn't quite get solved. Hooking up the replication streams to the sync protocol results in the hyperdrive stream never emitting an `end` event, stalling everything - in fact, not all events are sent properly.

I've spent a large chunk of the afternoon reconstructing this, approach, with some success, albeit no satisfying explanation. It's fine for the replication stream to pipe to a `Transform` stream - not a `Writable` stream - that doesn't produce any data but calls a callback. And if this callback (eventually) pushes data to another `Transform` stream - not a `Readable` - that pipes back to the replication stream, things work fine.

Which means the synchronization protocol needs to be refactored to use two helper `Transform` streams for dealing with the callbacks, and things should be good.

**Comment by jfinkhaeuser** [comment]
So far, so promising... the solution pushed can certainly replicate repos by way of protocols (tested, now also with hyperdrive!). Hooking this up with the sync protocol is WIP.

**Comment by jfinkhaeuser** [comment]
Synchronizing a single repo over the network works, so now the only WIP left is doing this with multiple repos.

**Comment by jfinkhaeuser** [comment]
Due to the decisions going into https://github.com/orgs/Joystream/projects/7 I think we can close this issue and the associated PR.

This doesn't mean there's a fix, but Acropolis will take a different approach.
